sasl/scram/scram.go

```diff
@@ -0,0 +1,83 @@
+package scram
+
+import (
+	"context"
+	"crypto/sha512"
+	"hash"
+
+	"github.com/pkg/errors"
+	"github.com/xdg/scram"
+)
+
+// HashFunction determines the hash function used by SCRAM to protect the user's
+// credentials.
+type HashFunction int
```

Comment:
Any reasons not to make this an interface with implementations for the SHA256 and SHA512 hash functions instead of an enum type?

Reply:
I did have reasons...the most important of which was to make sure we didn't export anything 3rd party libraries (e.g. in case we swap out scram providers later).  But I think that can also be done with an interface.  I just pushed up what I would see as the interface version...let me know which you prefer!